src/composeappmanager.cc

```diff
@@ -145,25 +227,274 @@ data::InstallationResult ComposeAppManager::install(const Uptane::Target &target
 //  2) update is applied, so we are now running both app1 and app2
 //  3) sota.toml is updated with 1 docker app: "app1"
 // At this point we should stop app2 and remove it.
-void ComposeAppManager::handleRemovedApps(const Uptane::Target &target) const {
+void ComposeAppManager::handleRemovedApps(const Uptane::Target& target) const {
   if (!boost::filesystem::is_directory(cfg_.apps_root)) {
     LOG_DEBUG << "cfg_.apps_root does not exist";
     return;
   }
   std::vector<std::string> target_apps = target.custom_data()["docker_compose_apps"].getMemberNames();
 
-  for (auto &entry : boost::make_iterator_range(boost::filesystem::directory_iterator(cfg_.apps_root), {})) {
+  for (auto& entry : boost::make_iterator_range(boost::filesystem::directory_iterator(cfg_.apps_root), {})) {
     if (boost::filesystem::is_directory(entry)) {
       std::string name = entry.path().filename().native();
       if (std::find(cfg_.apps.begin(), cfg_.apps.end(), name) == cfg_.apps.end()) {
         LOG_WARNING << "Docker Compose App(" << name
                     << ") installed, but is now removed from configuration. Removing from system";
-        ComposeApp(name, cfg_).remove();
+        ComposeApp(name, cfg_, registry_client_).remove();
       } else if (std::find(target_apps.begin(), target_apps.end(), name) == target_apps.end()) {
         LOG_WARNING << "Docker Compose App(" << name
                     << ") configured, but not defined in installation target. Removing from system";
-        ComposeApp(name, cfg_).remove();
+        ComposeApp(name, cfg_, registry_client_).remove();
       }
     }
   }
 }
+
+std::pair<std::string, std::string> ComposeApp::parseAppUri(const std::string& uri) {
+  auto split_pos = uri.find('@');
+  if (split_pos == std::string::npos) {
+    throw std::invalid_argument("Invalid App URI: '@' not found in " + uri);
+  }
+
```

Comment:
Can't it be different than `hub.foundries.io`, e.g. test mocks, tests, on prem deployment. I suppose it can be verified/checked if it's inline with what is defined in the config (registry_base_url).

Reply:
this all fits in with my other comment above about just letting docker's api tell us what to authenticate with.

Reply:
Maybe I just check if the App URI hostname matches the registry_base_url's hostname here, and check if the registry_base_url is equal to `https://hub.foundries.io` at the configuration time. 
Another option is to get rid off `base_url`, auth_token_endpoint, repo_base_url configuration variables at all and extracts registry hostname from App URI, although '/token-auth/' and /v2/ has to be hardcoded then. BTW, /token-auth/ is hardcoded in the docker compose download implementation,  too.

Reply:
I think getting rid of them is fine for now. "v2" can be hard-coded. I guess I did "auth-token" incorrectly in my docker-compose patch. The "Www-Authenticate:" header should tell the client where to go. For docker, I noticed my logic would be broke as the do:
~~~
Www-Authenticate: Bearer realm="https://auth.docker.io/token",service="registry.docker.io"...
~~~